fil_fungible_token/src/runtime/messaging.rs

```diff
@@ -0,0 +1,160 @@
+    fn call_receiver_hook(
+        &self,
+        from: ActorID,
+        to: ActorID,
+        token_value: &TokenAmount,
+        data: &[u8],
+    ) -> Result<Receipt>;
+
+    /// Resolves the given address to it's ID address form
+    fn resolve_id(&self, address: &Address) -> Result<ActorID>;
+
+    ///  Creates an account at a pubkey address and returns the ID address
+    fn initialise_account(&self, address: &Address) -> Result<ActorID>;
+}
+
+#[derive(Debug, Default, Clone, Copy)]
+pub struct FvmMessenger {}
+
+impl Messaging for FvmMessenger {
+    fn call_receiver_hook(
+        &self,
+        from: ActorID,
+        to: ActorID,
+        token_value: &TokenAmount,
+        data: &[u8],
+    ) -> Result<Receipt> {
+        // TODO: use fvm_dispatch here (when it supports compile time method resolution)
+        // TODO: ^^ necessitates determining conventional method names for receiver hooks
+
+        // currently, the method number comes from taking the name as "TokensReceived" and applying
+        // the transformation described in https://github.com/filecoin-project/FIPs/pull/399
+        const METHOD_NUM: u64 = 1361519036;
+        let to = Address::new_id(to);
+
+        let params = TokenReceivedParams {
+            sender: Address::new_id(from),
+            value: token_value.clone(),
+            data: RawBytes::from(data.to_vec()),
+        };
+        let params = RawBytes::new(fvm_ipld_encoding::to_vec(&params)?);
+
+        Ok(send::send(&to, METHOD_NUM, params, TokenAmount::zero())?)
+    }
+
+    fn resolve_id(&self, address: &Address) -> Result<ActorID> {
+        actor::resolve_address(address).ok_or(MessagingError::AddressNotInitialised(*address))
+    }
+
+    fn initialise_account(&self, address: &Address) -> Result<ActorID> {
+        if let Err(e) = send::send(address, METHOD_SEND, Default::default(), TokenAmount::zero()) {
+            return Err(e.into());
+        }
+
+        actor::resolve_address(address).ok_or(MessagingError::AddressNotInitialised(*address))
+    }
+}
+
+/// A fake method caller
+///
+/// # call_receiver_hook
+/// If call_receiver_hook is called with an empty data array, it will return true.
+/// If call_receiver_hook is called with a non-empty data array, it will return false.
+///
+/// # resolve_id
+/// If Address is id-address or secp-address, it resolves by returing FAKE_RESOLVED_ID
+/// If Address is actor-address or bls-address, it returns an error simulating an uninitialised address
+///
+/// # initialise_account
+/// Simulates the initialisation of an account
+/// - ID, SECP addresses panic as they should not be called given that resolve_id gave back an ActorID
+/// - BLS addresses give back FAKE_INITIALIZED_ID
+/// - Actor addresses are uninitialised and give back an error
+#[derive(Debug, Default, Clone, Copy)]
+pub struct FakeMessenger {}
+
+pub const FAKE_RESOLVED_ID: ActorID = 100;
+pub const FAKE_INITIALIZED_ID: ActorID = 101;
```

Comment:
Probably not an issue right now, but really these should be variables so we can return distinct ids for different addresses. In fact they should be a map of already resolved ones so they are stable. This will probably bite someone at some point.  You can defer to #33 if you like.